compiler/optimizer/abstractinterpreter/AbsLocalVarArray.hpp

```diff
@@ -0,0 +1,82 @@
+
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+
+#ifndef ABS_LOCAL_VAR_ARRAY_INCL
+#define ABS_LOCAL_VAR_ARRAY_INCL
+
+#include "infra/deque.hpp"
+#include "env/Region.hpp"
+#include "optimizer/ValuePropagation.hpp"
+#include "optimizer/abstractinterpreter/AbsValue.hpp"
+
+/**
+.* Local Variable Array Simulation.
+ */
+class AbsLocalVarArray
+   {
+   public:
+   AbsLocalVarArray(TR::Region &region);
+   AbsLocalVarArray(AbsLocalVarArray&, TR::Region& region);
```

Comment:
This is a copy constructor for AbsLocalVarArray

Reply:
Although this constructor does effectively copy an object, it is in fact not a [copy constructor](https://en.cppreference.com/w/cpp/language/copy_constructor), hence my question about the use case. All the constructor appears to do is create a copy of an existing object (by reference, which is a bit unusual) but with a different `TR::Region` as allocator. Is that really such a common operation?

Reply:
This is used during copying an AbsState

Reply:
Ok, that part makes sense, but what about the `TR::Region*`? Is the intention to make the copy use a different `TR::Region`?

Also, if this is just doing a copy, the original object should be received as a const ref.

Reply:
Because TR::deque requires a TR::Region& and AbsLocalVarArray itself does not keep the region where it is allocated.

Reply:
So why doesn't `AbsLocalVarArray` keep track of the region?

I'm trying to understand how this constructor is supposed to be used and I'm not sure I understand when/why the allocator would be different for the copy.

Reply:
The allocator is being passed from the copy constructor of `AbsState`. Because we may copy the `AbsState` on a different region (i.e. stackRegion and heapRegion). For example, we may copy an `AbsState` just for temporary use, so we copy it on a stackRegion. For life long usage, we may need to allocate it on a heapRegion.